File: backend/src/controllers/chat.controller.js
const Message = require('../models/Message');
const User = require('../models/User');

async function canCommunicate(user, partnerId) {
  const partner = await User.findById(partnerId);
  if (!partner) return false;

  if (user.role === 'patient' && partner.role === 'doctor') {
    return String(user.assignedDoctor) === String(partner._id);
  }
  if (user.role === 'doctor' && partner.role === 'patient') {
    return String(partner.assignedDoctor) === String(user._id);
  }
  if (user.role === 'admin') return true;
  return false;
}

exports.getConversation = async (req, res) => {
  try {
    const { partnerId } = req.params;
    const conversationId = Message.buildConversationId(req.user._id, partnerId);

    const messages = await Message.find({ conversationId })
      .sort({ createdAt: 1 })
      .limit(200);

    res.json({
      conversationId,
      encrypted: true,
      messages: messages.map((m) => m.toClient()),
    });
  } catch (error) {
    res.status(500).json({ message: 'Failed to load conversation' });
  }
};

exports.getPartners = async (req, res) => {
  try {
    if (req.user.role === 'patient') {
      if (!req.user.assignedDoctor) {
        return res.json({ partners: [] });
      }
      const doctor = await User.findById(req.user.assignedDoctor).select(
        'firstName lastName specialty role email'
      );
      return res.json({ partners: doctor ? [doctor] : [] });
    }

    if (req.user.role === 'doctor') {
      const patients = await User.find({
        role: 'patient',
        assignedDoctor: req.user._id,
      }).select('firstName lastName email role');
      return res.json({ partners: patients });
    }

    if (req.user.role === 'admin') {
      const partners = await User.find({
        _id: { $ne: req.user._id },
        isActive: true,
        role: { $in: ['doctor', 'patient'] },
      }).select('firstName lastName email role specialty');
      return res.json({ partners });
    }

    res.json({ partners: [] });
  } catch (error) {
    res.status(500).json({ message: 'Failed to load chat partners' });
  }
};

/** REST send — works on Vercel serverless (Socket.io optional locally). */
exports.sendMessage = async (req, res) => {
  try {
    const { receiverId, content } = req.body;
    if (!receiverId || !content?.trim()) {
      return res.status(400).json({ message: 'receiverId and content are required' });
    }

    const allowed = await canCommunicate(req.user, receiverId);
    if (!allowed) {
      return res.status(403).json({ message: 'Not allowed to message this user' });
    }

    const msg = await Message.createEncrypted({
      sender: req.user._id,
      receiver: receiverId,
      content: content.trim(),
    });

    res.status(201).json({ message: msg.toClient() });
  } catch (error) {
    console.error('sendMessage:', error);
    res.status(500).json({ message: 'Failed to send message' });
  }
};

File: backend-dotnet/MyHeath.Api/Controllers/ChatController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using MongoDB.Driver;
using MyHeath.Api.Infrastructure;
using MyHeath.Api.Models;
using MyHeath.Api.Services;

namespace MyHeath.Api.Controllers;

[ApiController]
[Route("api/chat")]
[Authorize]
public class ChatController : ApiControllerBase
{
    private readonly AesCryptoService _aes;

    public ChatController(MongoContext db, AesCryptoService aes) : base(db) => _aes = aes;

    public record SendDto(string ReceiverId, string Content);

    [HttpGet("partners")]
    public async Task<IActionResult> Partners()
    {
        var user = await CurrentUserAsync();
        if (user is null) return Fail("Unauthorized", 401);

        if (user.Role == Roles.Patient)
        {
            if (user.AssignedDoctor is null) return Ok(new { partners = Array.Empty<object>() });
            var doctor = await Db.Users.Find(u => u.Id == user.AssignedDoctor).FirstOrDefaultAsync();
            return Ok(new
            {
                partners = doctor is null ? Array.Empty<object>() : new object[]
                {
                    new { _id = doctor.Id, id = doctor.Id, firstName = doctor.FirstName, lastName = doctor.LastName,
                        specialty = doctor.Specialty, role = doctor.Role, email = doctor.Email }
                }
            });
        }

        if (user.Role == Roles.Doctor)
        {
            var patients = await Db.Users.Find(u => u.Role == Roles.Patient && u.AssignedDoctor == user.Id).ToListAsync();
            return Ok(new
            {
                partners = patients.Select(p => new
                {
                    _id = p.Id, id = p.Id, firstName = p.FirstName, lastName = p.LastName,
                    email = p.Email, role = p.Role
                })
            });
        }

        if (user.Role == Roles.Admin)
        {
            var partners = await Db.Users.Find(u => u.Id != user.Id && u.IsActive &&
                (u.Role == Roles.Doctor || u.Role == Roles.Patient)).ToListAsync();
            return Ok(new
            {
                partners = partners.Select(p => new
                {
                    _id = p.Id, id = p.Id, firstName = p.FirstName, lastName = p.LastName,
                    email = p.Email, role = p.Role, specialty = p.Specialty
                })
            });
        }

        return Ok(new { partners = Array.Empty<object>() });
    }

    [HttpPost("send")]
    public async Task<IActionResult> Send([FromBody] SendDto dto)
    {
        var user = await CurrentUserAsync();
        if (user is null) return Fail("Unauthorized", 401);
        if (string.IsNullOrWhiteSpace(dto.ReceiverId) || string.IsNullOrWhiteSpace(dto.Content))
            return Fail("receiverId and content are required");

        if (!await CanCommunicate(user, dto.ReceiverId))
            return Fail("Not allowed to message this user", 403);

        var msg = new Message
        {
            Sender = user.Id,
            Receiver = dto.ReceiverId,
            EncryptedContent = _aes.Encrypt(dto.Content.Trim()),
            ConversationId = BuildConversationId(user.Id, dto.ReceiverId)
        };
        await Db.Messages.InsertOneAsync(msg);
        return StatusCode(201, new { message = ToClient(msg) });
    }

    [HttpGet("{partnerId}")]
    public async Task<IActionResult> Conversation(string partnerId)
    {
        var user = await CurrentUserAsync();
        if (user is null) return Fail("Unauthorized", 401);
        var cid = BuildConversationId(user.Id, partnerId);
        var messages = await Db.Messages.Find(m => m.ConversationId == cid)
            .SortBy(m => m.CreatedAt).Limit(200).ToListAsync();
        return Ok(new
        {
            conversationId = cid,
            encrypted = true,
            messages = messages.Select(ToClient)
        });
    }

    private async Task<bool> CanCommunicate(User user, string partnerId)
    {
        var partner = await Db.Users.Find(u => u.Id == partnerId).FirstOrDefaultAsync();
        if (partner is null) return false;
        if (user.Role == Roles.Admin) return true;
        if (user.Role == Roles.Patient && partner.Role == Roles.Doctor)
            return user.AssignedDoctor == partner.Id;
        if (user.Role == Roles.Doctor && partner.Role == Roles.Patient)
            return partner.AssignedDoctor == user.Id;
        return false;
    }

    private static string BuildConversationId(string a, string b)
    {
        var pair = new[] { a, b }.OrderBy(x => x, StringComparer.Ordinal).ToArray();
        return $"{pair[0]}_{pair[1]}";
    }

    private object ToClient(Message m) => new
    {
        id = m.Id,
        _id = m.Id,
        sender = m.Sender,
        receiver = m.Receiver,
        content = _aes.DecryptToString(m.EncryptedContent),
        conversationId = m.ConversationId,
        read = m.Read,
        createdAt = m.CreatedAt
    };
}
